src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -412,13 +447,13 @@ public NodeStateTable<T> ToImmutableAndFree()
                     hasTrackedSteps: TrackIncrementalSteps);
             }
 
-            private static (T chosen, EntryState state) GetModifiedItemAndState(T previous, T replacement, IEqualityComparer<T> comparer)
+            private static (T chosen, EntryState state, bool chosePrevious) GetModifiedItemAndState(T previous, T replacement, IEqualityComparer<T> comparer)
             {
                 // when comparing an item to check if its modified we explicitly cache the *previous* item in the case where its 
                 // considered to be equal. This ensures that subsequent comparisons are stable across future generation passes.
                 return comparer.Equals(previous, replacement)
-                    ? (previous, EntryState.Cached)
-                    : (replacement, EntryState.Modified);
+                    ? (previous, EntryState.Cached, chosePrevious: true)
+                    : (replacement, EntryState.Modified, chosePrevious: false);
```

Comment:
note: chosePrevious is technically redundant.  we can tell waht happened by seeing if we got `EntryState.Cached` or `EntryState.Modified` back.  But i was worreid that was too confusing/brittle, so i went this path.